pkg/resources/credentials.go

```diff
@@ -282,32 +284,45 @@ func GetOpenstackCredentials(data CredentialsData) (OpenstackCredentials, error)
 	openstackCredentials := OpenstackCredentials{}
 	var err error
 
-	if spec.Username != "" {
-		openstackCredentials.Username = spec.Username
-	} else if openstackCredentials.Username, err = data.GetGlobalSecretKeySelectorValue(spec.CredentialsReference, OpenstackUsername); err != nil {
-		return OpenstackCredentials{}, err
+	if spec.ApplicationCredentialID != "" {
+		openstackCredentials.ApplicationCredentialID = spec.ApplicationCredentialID
+	} else {
+		openstackCredentials.ApplicationCredentialID, _ = data.GetGlobalSecretKeySelectorValue(spec.CredentialsReference, OpenstackApplicationCredentialID)
 	}
-
-	if spec.Password != "" {
-		openstackCredentials.Password = spec.Password
-	} else if openstackCredentials.Password, err = data.GetGlobalSecretKeySelectorValue(spec.CredentialsReference, OpenstackPassword); err != nil {
-		return OpenstackCredentials{}, err
+	if spec.ApplicationCredentialSecret != "" {
+		openstackCredentials.ApplicationCredentialSecret = spec.ApplicationCredentialSecret
+	} else {
+		openstackCredentials.ApplicationCredentialSecret, _ = data.GetGlobalSecretKeySelectorValue(spec.CredentialsReference, OpenstackApplicationCredentialSecret)
 	}
 
-	if spec.Tenant != "" {
-		openstackCredentials.Tenant = spec.Tenant
-	} else if spec.CredentialsReference != nil && spec.CredentialsReference.Name != "" {
-		if openstackCredentials.Tenant, err = data.GetGlobalSecretKeySelectorValue(spec.CredentialsReference, OpenstackTenant); err != nil {
+	if openstackCredentials.ApplicationCredentialID == "" || openstackCredentials.ApplicationCredentialSecret == "" {
```

Comment:
```suggestion
	if openstackCredentials.ApplicationCredentialID == "" && openstackCredentials.ApplicationCredentialSecret == "" {
```

I think here we should lookup for login/pass only if both are empty. Because if only 1 is empty and other is not it means there was a try to provide them, but something went wrong.

Reply:
I agree.